api/src/server.rs

```diff
@@ -399,27 +407,45 @@ where
             // rely on the channel depth for backpressure. the goal is to keep the queue close to empty
             // without processing one at a time. when we stop parsing the carfile in the store
             // i.e. validate before sending here and this is just an insert, we may want to process more at once.
+            let mut shutdown = false;
             loop {
                 let mut buf = Vec::with_capacity(EVENTS_TO_RECEIVE);
+                let mut process_early = false;
                 tokio::select! {
                     _ = interval.tick() => {
-                        Self::process_events(&mut events, &event_store, node_id).await;
+                        process_early = true;
                     }
                     val = event_rx.recv_many(&mut buf, EVENTS_TO_RECEIVE) => {
                         if val > 0 {
                             events.extend(buf);
                         }
                     }
-                }
-                let shutdown = event_rx.is_closed();
-                // make sure the events queue doesn't get too deep when we're under heavy load
-                if events.len() >= EVENT_INSERT_QUEUE_SIZE || shutdown {
-                    Self::process_events(&mut events, &event_store, node_id).await;
-                }
+                    _ = shutdown_signal.recv() => {
+                        tracing::debug!("Insert many task got shutdown signal");
+                        shutdown = true;
+                    }
+                };
                 if shutdown {
-                    tracing::info!("Shutting down insert task.");
+                    tracing::info!("Shutting down insert task after processing current batch");
+                    if !event_rx.is_empty() {
+                        let remaining_event_cnt = event_rx.len();
+                        let mut buf = Vec::with_capacity(remaining_event_cnt);
```

Comment:
can we not just use the buf that was already allocated on line 412?

Reply:
Yeah, I makes sense. We can resize and use it (or just use it since recv_many might resize automatically - I’ll verify). It’s the last allocation before exiting so I didn’t think too closely about it.

Reply:
We can't use it actually. it gets moved into the select branch even though we don't follow that arm in this case. I added a comment to that effect. expanding the macro might make it easier to see but I didn't bother.